### ISSUE TITLE
Add a Spreadsheet resource for handling tabular data.

### ISSUE THREAD

**Issue by willnationsdev** [godotengine/godot-proposals#13]
**Do you need this for a project you are working on? If so, please describe your project and how this will help you.**

Yes, I have a project that will likely involve several different in-game abilities with varying stats associated with their effects. It is common for me to use a tabular editor (like Excel) to model and iterate data of this sort. Being able to edit spreadsheets that are integrated with Godot's scripting API will assist me in developing and maintaining a more type-safe data structure for my project.

 **Is this something that you will need to use often, or which can’t be easily worked around? Why or why not?**

I generally like games that involve a lot of data on the backend, so I foresee myself (and others, who I know are seeking to use CSV files in their projects) using this often. You could work around it with `File.get_csv_line()`, but it several problematic differences would exist:

1. There would be no GUI for easily editing the data directly from the Godot editor for non-tech-savvy individuals (who may become part of my project down the line - I want to leave things tinkerable without them needing to open a separate application just to edit some data for the game).
2. It wouldn't be type-safe. A spreadsheet editing experience in Godot could specify a Script model to map to the columns of the editor, thereby guaranteeing the type-safety of the fields in the data.
3. As the EditorInspector already knows how to render custom editors for specific types of variables, the editing experience for any individual property is specially tailored for editing that type of data, making it more intuitive for non-programmers. A similar editing experience can not be found in a text editor for a CSV file or from within a significantly more complex, context-unaware application like Excel.
4. I would like to be able to leverage property magic methods (`_get`, `_set`, `_get_property_list`) to manipulate how different kinds of resources are rendered and the EditorInspector already fully supports this. This is not as easy to implement out-of-the-box in Excel where logic would need to be re-implemented or in a text editor for a CSV file which...has no GUI.

 **Are you able to submit a pull request if this proposal is approved?**

Yes.

 **Please describe the proposal:**

I propose adding a Spreadsheet resource that is specifically designed for editing tabular data.

In addition to the benefits mentioned above, it would also easily fill the role of properly importing CSV/TSV files as data structures rather than exclusively as translation files. This would overall enable more options for designers using Godot as well as make interop with other software applications easier since CSV is a very common format, similar to JSON.

 **Please describe the use case with a mock-up, video or diagram of examples on how it will be used:**

This concept is very similar to the UE4 DataTable asset type, which I drew pieces from in order to create the mockup. In this example, there is a "skill_data.gd" resource script that is constraining the resource instances and consequent editors available for the fields in the spreadsheet editor. You use the bottom panel to view/add/delete/select resources and use the Inspector to edit the fields of those resources.

![spreadsheet_mockup](https://user-images.githubusercontent.com/16217563/64219253-6366c900-ce8a-11e9-9181-b0bd16ec995f.png)

 **Optionally, please provide any technical implementation suggestions you have. They may or may not be part of the final implementation if approved:**

This would largely involve creating the Spreadsheet resource in a module, adding a ResourceImporter for CSV/TSV files that can convert to Spreadsheet resources, and then creating the bottom panel editor.

Technically speaking, there is nothing that stops all of this from just being implemented in a plugin through the scripting API, but I'm confident that this is something that would be popular enough that people would want the functionality built into the engine rather than having to install a plugin to use it, similar to how JSON is integrated in some capacity.

**Comment by golddotasksquestions** [comment]
I would greatly appreciate not having to write a parser for spreadsheets but native CSV support that would already parse the file on import, and provide it to me as Array or Dictionary.
This wound be a feature used by game designers (who are not necessarily programmers), so UI/UX should 
Usecases: Any game design that requires data driven balancing: Management or strategy games, FPS with a wide range of weapons, cardgames (every row resembles a card), ...
Currently I am working on a cardgame that would need this feature.

**Comment by willnationsdev** [comment]
@golddotasksquestions Note that, in this example, a CSV would be imported not as a Dictionary or Array, but as a Spreadsheet which wraps the data and would provide various means of accessing it depending on the settings it has been configured with

For example, the `uses_keys` bool property would convert the interface between a Dictionary and an Array of Arrays, depending on whether the spreadsheet has been configured to understand how to interpret a row's primary key (in database terms). This is similar to my GDScript-based CSVFile implementation in [godot-next](https://github.com/willnationsdev/godot-next/tree/master/addons/godot-next/references/csv_file.gd) (which you know because you've tinkered with it, but I'm stating for context purposes).

**Comment by willnationsdev** [comment]
I suppose it would be better to implement a full proof of concept in GDScript using an EditorPlugin since...as far as I can tell, all of the features that would be required can be done through an EditorPlugin.

**Comment by lentsius-bark** [comment]
I make extensive use of tabular data in in two of my projects and would 100% welcome this.

**Comment by Zylann** [comment]
If there was such resource type, one could then write a plugin to support importing ODS or XLSX spreadsheets :p

**Comment by willnationsdev** [comment]
Can't say I haven't thought about also supporting ODS/XLSX to some extent, but I think that once you've gotten to that point, it just goes beyond what is really needed by a game engine. Being able to properly import a CSV as its own spreadsheet resource is as much as we need, and it's a small enough in scope that I think it warrants a place in the main engine repository (doesn't really need to be in the engine core itself - stick mostly to module/editor stuff).

**Comment by Zylann** [comment]
When I said "core" I meant "built-in", sorry :)

**Comment by willnationsdev** [comment]
@Zylann Yeah, we really need to decide on official terms to differentiate stuff like that. Currently there's...

- the "engine core", i.e. the /core subdirectory.
- the "core engine", i.e. the parts of the codebase the constitute the actual engine itself:
    - /core
    - /main
    - /scene
    - /servers
    - possibly parts of /drivers or /platform, although those are SUPPOSED to be separate dependencies. I know for a fact I once tried to separate out the /core/map.h file from the rest of the engine to use it in a GDNative project and low-and-behold it eventually twisted its dependencies into the OS class and from there touched literally everything else in platform and other parts of the engine. -_-
    - This DEFINITELY does not include /editor and /modules. Probably doesn't include /thirdparty, though I can't be sure since I don't know what all in there is used where.
- The main engine repository versus external plugins / gdnative plugins, and external engine modules which are kept in separate repositories.

**Comment by golddotasksquestions** [comment]
Spreadsheets are so essential to gamedesign and the profession of a game designer, it sounds quite absurd to me if my engine of choice would only interpret them as translation files or would need a plugin to access it's data.

**Comment by Zylann** [comment]
Keep in mind there is absolutely no problem at all loading such data. You can load CSV in a few lines of scripts and write specific editors with plugins too. But a spreadsheet can be interpreted in numerous ways, that's where making such a thing "built-in" becomes difficult because the engine cannot guess what users will do with this. Even translations as CSV is borderline to me, because it forces users to have a convention imposed by the engine. But for game-specific data? There si no convention here.

**Comment by willnationsdev** [comment]
@golddotasksquestions The File class actually has `get_csv_line()` and `store_csv_line()` methods now which allow you to convert PoolStringArrays into CSV rows. The functionality of parsing and using them is all built-in already and allows one to use it however they want. The 3.2 release is also going to have a [new CSV importer](https://github.com/godotengine/godot/pull/31927) (not the default, unfortunately), which should simplify the process of keeping CSV files in one's project, parsing them, and editing data inside of them.

@Zylann The problem I have is wanting to edit tabular data in a type-safe way, where the rows are guaranteed to follow a particular schema of data types dictated by a script I've defined. There is no proper editor for handling that. So, this goes beyond "[allowing] plugins to define extra import modes to CSV". It's about having a tabular data editor for Godot-defined data types.

The point of conflict here is whether the use of this editor would be wide-spread enough to warrant it being in the main engine repository since the vast majority of the work for this can be done within a plugin. And I plan on creating said plugin [here](https://github.com/willnationsdev/godot-spreadsheet) as a proof-of-concept. If we can show how useful it is / how many people star it / use it, then that should help indicate how valuable it could be as a built-in tool for the editor itself rather than having to be installed via a plugin.

Of course, another thing that could be done is relegating this to a plugin only and then adding a better system for more easily generating projects with a variety of plugins that one always wishes to use, etc. That's a little bit more like #62 though (but without the build-system part).

**Comment by golddotasksquestions** [comment]
CSVs are textfiles formated in a specific way. If I want to use them as text files to do whatever, I already can with File.new.
A CSV importer however could already interpret those as data sets, convert them to either and Array of Arrays or a Dictionary of Arrays, or translation. In the import setting there could be checkboxes for either, as well as the option to define if the first row or the first column should be the header.

**Comment by willnationsdev** [comment]
@golddotasksquestions I mean, if you really just want to simplify that workflow, then we could always write a CSV class, like JSON, that essentially does the same thing that File does but on the backend, converting a file to an Array of PoolStringArrays. That's a different request than this Issue though.

**Comment by nhydock** [comment]
Would we want the tabular data to remain only simple data types, or should each row be as flexible as a custom resource class and allow embedding of complex types such as vectors, curves, etc.?  If it's already wrapped as a Custom "Spreadsheet" Resource type instead of actually saving as .csv files, I figure the row definition should be allowed as any Resource class.

**Comment by willnationsdev** [comment]
@nhydock Having a Resource script used as a schema for defining the columns of each row is exactly what I had in mind, that way you would get type-safe access to all Godot types.

**Comment by willnationsdev** [comment]
I recently saw reduz's [tweet]() about typed arrays and the plans to eventually also have typed Dictionaries. If support for that were added, alongside support for user-defined resource type recognition, then I think we'd be well on our way to creating the backend for this since, at that point, it's just a matter of wrapping the data in an updated user interface and maybe having a sheets abstraction for any given Spreadsheet resource.

**Comment by aaronfranke** [comment]
Maybe you should make an editor plugin as a proof of concept, or as a complete solution. I don't really see why this needs to be in core, I have never had a use for this in Godot, so it sounds like a perfect candidate for an editor plugin instead of an engine feature.

**Comment by dragonDScript** [comment]
A file.get_csv_column() would be very useful too.

**Comment by dragonDScript** [comment]
In the core engine I mean

**Comment by Calinou** [comment]
@dragonDScript Please open a separate proposal for that.

**Comment by dragonDScript** [comment]
Sure I'll do, thank you.

**Comment by Xrayez** [comment]
After revisiting old proposals, I think I've unknowingly implemented a class in [Goost](https://goostengine.github.io/) which shares a lot of similarities regarding proposed `Spreadsheet` resource here.

See [`VariantMap`](https://goost.readthedocs.io/en/latest/classes/class_variantmap.html) class. Basically, it's a simple two-dimensional data container that map `Variant` types onto a grid. You can have any type stored in the grid at run-time with `set_cell()` and access the elements with `get_cell()`.

Unfortunately, Goost does not currently provide any kind of editors for the class to be useful for use cases such as editing spreadsheets etc, but in theory it should be quite possible to do. Since Goost implemented as a C++ module, it should be possible to reuse Godot's property editors to make it work in a general-purpose way, without reinventing the same property editors via script (which I believe would be a major blocker/limitation to implement this feature easily via script).

I've chosen `VariantMap` name because:
- it handles `Variant`s internally.
- it maps `Variant` values onto a grid.
- not ambiguous with UI `Control` classes (like proposed in #2662).
- familiar with classes like:
   - `BitMap` (which is basically the same thing, but only handles `bool` values);
   - `TileMap` (you can use `VariantMap` to conveniently store tile metadata alongside, which also seems to be a more or less common feature request: godotengine/godot#31529, godotengine/godot#12634).

**Comment by mia-bentzen** [comment]
I'm quite new to contributing to open source, so apologies if this is the wrong place to say this, or the wrong way to say it.

I think a really useful feature to have in this kind of editor would be some amount of basic formula (i think that's what excel calls it?) support. So you can either set values directly, or (for types that support it) create an expression (that may reference other values in the spreadsheet) that calculates what the value should be. I imagine this would probably be accomplished by using GDScript to handle the expressions.

I personally would be satisfied if it was just regular numbers it supported, but you could probably make it work for vectors and colors too.

**Comment by yoskeleton** [comment]
That would be the best thing to manage all my .TRES files !

Its so slow to find, open and edit each .TRES files one by one

Big +1 for this proposals !

**Comment by Calinou** [comment]
@yoskeleton Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by don-tnowe** [comment]
I was looking for a way to edit many resources at once and looked for some plugin Didn't find anything... So made exactly that:

[](https://github.com/don-tnowe/godot-resources-as-sheets-plugin)

It uses folders of Resources instead of some special Table datatype, so game logic does not depend on it. It can show Resources in the Inspector and allows appending text or setting a property in several Resources at once.

Also has some cool tools to manipulate numbers, shift and darken colors, and chop Textures into several AtlasTextures.



Pic:
![pic](https://user-images.githubusercontent.com/67479453/191821740-93eea94f-8c8c-4e2b-b8b0-f8a7e9573c38.png)

**Comment by Zireael07** [comment]
Current CSV import (whether in 3.5 or 4.0) is pretty unwieldy.

This is a sample of CSV data for my space game:

> name, winchellX, winchellY, winchellZ, color, planets, multiple, comments, ra, dec, dist-Sol
Proxima Centauri, 2.8ly, -3.1ly, 0.1ly, red, yes, -, //7200 AU or 0.13ly away from Alpha
Alpha Centauri, 2.9ly, -3.1ly, 0.1ly, yellow, yes,
Tau Ceti, -3.4ly, 0.4ly, -11.4ly, yellow, yes,
Barnard's Star, 5.0ly, 3.0ly, 1.4ly, red, yes, , , 17.96h, 4.7, 5.95

The way to parse it involves having to remember which column is which index, instead of writing data["name"] or data["color"]...

Currently need to add another column, and honestly, I'll likely try the add-on mentioned above just for that reason.

**Comment by jordanlis** [comment]
Hi everyone. 
Personaly, I think it could be really usefull to have something related to database and handling data inside godot. I think it could be really usefull to have a data editor such as the one that don-tnowe showed. 

But at some point, what's really important is not editing this file, which could be done in any software like excel. 

The real matter is handling the CSV file easily in gdscript code. 
For example if you have a weapon and that you want you access its properties, you have to parse it and so on. 

One thing that could be great could be just to create and access to this database easily in gdscript with something like this : 

**1. Create a database :** 
database = create_database_from_csv_file("name_of_the_table1.csv", "name_of_the_table2.csv", "and_so_on.csv")

**2. access to database entries**
database.table1.line[1] #Allows to access to each lines of the database, each parameters, etc.


So maybe what we should focus on should be to handle in code a database, a table, entries and parameters.

But maybe ressources already allows it, not sure about it.

**Comment by Zireael07** [comment]
@jordanlis Access to line is easy. Accessing the correct cell/column is the more annoying part currently.

**Comment by Zireael07** [comment]
One resource per kind is what @don-tnowe does, too. That addon you linked is afaict pretty much the same as @don-tnowe except it has some data access/handling functions. And I can't find Kehom's source code, just the blog

**Comment by jordanlis** [comment]
> 



> One resource per kind is what @don-tnowe does, too. That addon you linked is afaict pretty much the same as @don-tnowe except it has some data access/handling functions. And I can't find Kehom's source code, just the blog

Yep sorry for the link. here's the addon link : Yep, sorry, here's the addon : https://github.com/Kehom/GodotAddonPack/tree/target-gd3.2/addons in "keh_gddb" folder. 

You're right, this is pretty much the main thing, which is good so multiple people ended up with the same solution. Maybe it could help use to decide between solution 1 and 2 in my previous message.

**Comment by ishax-kos** [comment]
> Can't say I haven't thought about also supporting ODS/XLSX to some extent, but I think that once you've gotten to that point, it just goes beyond what is really needed by a game engine. 

After watching one of sakurai's gamedev videos, im now convinced that supporting a spreadsheet format with formulas is necessary. At least insofar that you would be able to read cached formula values.

**Comment by jordanlis** [comment]
Formulas, I don't know but a database in some format is mandatory according to me. It's just almost the case for the translation part, but it lacks some features or it is not adapted perfeclty.

**Comment by ishax-kos** [comment]
Google appears to ship an API to read from online spreadsheets. It would need to be build time, and this would be amazing for teams. It would go a long way towards enabling a "single source of truth" for large projects.

**Comment by Patchcoat** [comment]
It seems like this one feature request encompasses multiple different specific considerations.

1. Tabular data is very useful, necessary even, for large sections of game development. Godot should be able to handle it.
2. Resources are how Godot stores data. We should interface with that and provide an in-engine and type-safe way to edit them.
3. We should improve GDScript support for csv files, one of the simplest forms of tabular data.
4. We should have have formulae available when editing the spreadsheet.
5. We should be able to interface with external tools.

For an internal tool, I believe we should read from and write to a csv file, and improve the GDScript support for csv along the way. This means users can't store their formulae in the saved file, they must bake them down to the results. It also means the user can't store their formulae in an ODS file on the side, because they risk that file going out of sync when the csv is edited in the engine.

However, we keep the interface simple which reduces the surface area for bugs and the effort required to maintain it while serving most of the developers who want tabular data. Once proper csv support is in, then we can worry about reading from ODS files and parsing formulae. I realize the temptation is there to skip straight to ODS, because there are more features, but I believe the simpler problem more people want should be tackled first.

**Comment by jordanlis** [comment]
Hi all, 

According to me, datatable should use resources in godot because it allows to handle all the features of the engine but in a database format. So I developped a small add-on which does the following things for my game : 
- It displays in godot all the resources files in a choosen folder
- Each resources is a line in the database
- Each columns is a value for one resources

It is fully compatible for my game and allows to modify more easily the values that I need instead of opening each resources in godot, which is a nightmare to handle all my items. 

![resources_datatable](https://github.com/user-attachments/assets/ac09e7db-21df-46ab-abf1-0ec9e126a569)

If someone wants to help, I could create folder in github in order to collaborate and add features to it together.

Hope it helps someone.

**Comment by don-tnowe** [comment]
Hey @jordanlis! Looks really nice. Does this cover any use cases that https://github.com/don-tnowe/godot-resources-as-sheets-does not?

**Comment by azur-wolve** [comment]
> [...] So I developped a small add-on which does the following things for my game :
>
> If someone wants to help, I could create folder in github in order to collaborate and add features to it together.
> 
> Hope it helps someone.

yes
being able to edit values on a spreadsheet table is pretty handy
but sounds more like a plugin

well done!
think gonna use it myself too

**Comment by jordanlis** [comment]
> Hey @jordanlis! Looks really nice. Does this cover any use cases that https://github.com/don-tnowe/godot-resources-as-sheets-plugin does not?

Yes it is, but since your plugin does not work with godot 4, I had to find another solution. Or is it me and you fixed it since the last time I tried to use it ? 

@azur-wolve I will do that in the following days

**Comment by GiggyMantis** [comment]
This is mostly unrelated but I have a use case where I need to display an editable spreadsheet to the player.

**Comment by BaconEggsRL** [comment]
Core support for importing .csv files, displaying tabular data and charts would be really handy.
Not sure if there's a better way, but currently looking into several addon tools for this:

https://github.com/timothyqiu/godot-csv-data-importer
https://github.com/sericaer/godot_table
https://github.com/fenix-hub/godot-engine.easy-charts
